### DISCUSSION TITLE
How to update NodeView attributes in a command?

### DISCUSSION THREAD

**Discussion by amirhhashemi** [ueberdosis/tiptap#2873] [General Questions]
I've been working on adding upload feature to the image extension. I think a command is the right place to do that so I wrote this command to upload the image to the server and update the image's src attribute with the new URL:
```tsx
  addCommands() {
    return {
      uploadImage:
        (dataUrl: string) =>
        async ({ commands }: CommandProps) => {
          const { url } = await upload(dataUrl);
          return commands.updateAttributes("image", { src: url })
        },
    };
  },
```
And I chain it with the `setImage` command (`setImage` just shows a temporary image to the user)
```tsx
    editor
      .chain()
      .focus()
      .setImage({ src: dataUrl })
      .uploadImage(dataUrl)
      .run();
```
But `uploadImage` doesn't update the node's attributes and I don't know why. worth mentioning that I use a NodeView to render the image:

```tsx
  addNodeView() {
    return ReactNodeViewRenderer(CustomImageComponent);
  },
```
CustomImageComponent
```tsx
export const CustomImageComponent = ({ node }: NodeViewProps) => {

  return (
    <NodeViewWrapper as="figure" data-type="custom-image">
      {node.attrs.src && (
        <>
          <img
            src={node.attrs.src}
            alt="Image"
            className={`${node.attrs.src.startsWith("data:") && "blur-sm"}`}
          />
          <NodeViewContent as="figcaption" className="text-center" />
        </>
      )}
    </NodeViewWrapper>
  );
};
```

**Comment by pointnet** [comment]
Is the image selected in the state before applying updateAttribute?

**Reply by amirhhashemi** [reply]
I run `uploadImage` right after `setImage` (with `chain()`). I'm not sure if it's selected or not but I don't move the cursor after I add the image. I've added more information in case you need it.

**Reply by hzpeng57** [reply]
@pointnet updateAttribute is executed correctly when the image is selected, but how to handle it when there are two images?

**Comment by ahukkanen** [comment]
I am experiencing the same issue with async command that would open a modal and then after the modal is saved, it would continue the command with the data received from the modal.

Here is a sample code to get the idea what I'm trying to do:

```js
import Youtube from "@tiptap/extension-youtube";

import InputModal from "./input_modal";

export default Youtube.extend({
  name: "video",

  addCommands() {
    return {
      ...this.parent?.(),
      videoModal: () => async ({ dispatch, commands }) => {
        if (dispatch) {
          const videoModal = new InputModal({ ... });
          const modalState = await videoModal.toggle();
          if (modalState !== "save") {
            return false;
          }
          commands.setYoutubeVideo({ width: 640, height: 480, src: videoModal.getValue("src") });
        }
        return true;
      }
    }
  }
});
```

Also trying to do the same thing with the image extension (using `commands.setImage({ src, alt })` with a similar approach) and the command does not work.

I have also tried `chain().focus().setYoutubeVideo({ ... }).run()` (as well as with the image) but it does not seem to work.

What I ended up doing to bypass this is to manually dispatching new transactions directly through ProseMirror and that seems to work properly. Would be nice if TipTap supported this, though, as the above example would be obviously much less code.

Using TipTap version 2.0.0-beta.209.

**Reply by ahukkanen** [reply]
Answering myself but the above command actually works if call `this.editor` instead of the editor object provided to the command as the first argument.

So this slight change to that code will make it work well.

Instead of:

```js
commands.setYoutubeVideo({ width: 640, height: 480, src: videoModal.getValue("src") });
```

Do:

```js
this.editor.commands.setYoutubeVideo({ width: 640, height: 480, src: videoModal.getValue("src") });
```

I guess this has something to do with the transaction context passed to the command argument(s).

**Reply by tyler-g** [reply]
Just noting here in case someone is wondering why the destructed `{commands}` or `{editor, commands}` in an `addCommands` callback could be a bad idea:

@ahukkanen the reason this works is because `commands` in this case is not a reference to the editor instance commands, but rather a destructured instance of it (basically meaning a copy).  

 It's really easy to get caught up on this and I've spent countless debugging hours on this oversight.

Long story short, it's generally a bad idea to use destructured anything related to the tip-tap editor if you are doing anything other than read-only operations for this reason.  That's why `this.editor` works in your case

**Reply by ahukkanen** [reply]
I think it would be extremely useful if such async command example was added somewhere in the documentation.

The documentation is always suggesting to use the destructed copies instead of the editor instance when dealing with the command execution. It was pretty much trial and error to figure that out.
